src/plugins/hidsim/gui/vmi_win_gui_parser.cpp

```diff
@@ -0,0 +1,1014 @@
+    GHashTable* seen_windows = g_hash_table_new_full(g_int64_hash, g_int64_equal,
+            free_data, NULL);
+
+    status_t ret = traverse_windows_pid(vmi, spwnd, pid, seen_windows, *result_windows, 0);
+
+    g_hash_table_destroy(seen_windows);
+
+    return ret;
+}
+
+/* Traverses this singly-linked list of desktops belonging to one WinSta */
+status_t traverse_desktops(vmi_instance_t vmi, addr_t* desktops,
+    size_t* max_len, addr_t list_head)
+{
+    addr_t cur = list_head;
+    addr_t next = 0;
+    size_t i = 0;
+
+    while (cur)
+    {
+        if (i < *max_len - 1)
+        {
+            desktops[i] = cur;
+            i++;
+        }
+        if (VMI_FAILURE == vmi_read_addr_va(vmi, cur + symbol_offsets.desk_rpdesk_next_off, 0, &next))
+        {
+            printf("Failed to read pointer to next desktop at %" PRIx64 "\n",
+                cur + symbol_offsets.desk_rpdesk_next_off);
+            *max_len = i;
+            return VMI_FAILURE;
+        }
+
+        if (next == list_head)
+            break;
+        cur = next;
+    }
+    *max_len = i;
+
+    return VMI_SUCCESS;
+}
+
+/*
+ * Reads relevant data from tagWINDOWSTATION-structs and the children of
+ * type tagDESKTOP
+ */
+status_t populate_winsta(vmi_instance_t vmi, struct winsta* winsta, addr_t addr,
+    vmi_pid_t providing_pid)
+{
+    winsta->addr = addr;
+
+    /*
+     * Do it like volatility: Find a process with matching sessionID and take its
+     * VA as _MM_SESSION_SPACE
+     */
+    winsta->providing_pid = providing_pid;
+
+    /* Reads pointer to global atom table */
+    if (VMI_FAILURE == vmi_read_addr_va(vmi, addr + symbol_offsets.winsta_pglobal_atom_table_offset, 0, &winsta->atom_table))
+    {
+        printf("Failed to read pointer to atom table at %" PRIx64 "\n", addr +
+            symbol_offsets.winsta_pglobal_atom_table_offset);
+        return VMI_FAILURE;
+    }
+
+    if (VMI_FAILURE == vmi_read_32_va(vmi, addr + symbol_offsets.winsta_session_id_offset, 0, &winsta->session_id))
+    {
+        printf("Failed to read session ID at %" PRIx64 "\n", addr + symbol_offsets.winsta_session_id_offset);
+        return VMI_FAILURE;
+    }
+
+    uint32_t wsf_flags = 0;
+
+    if (VMI_FAILURE == vmi_read_32_va(vmi, addr + symbol_offsets.winsta_wsf_flags, 0, &wsf_flags))
+    {
+        printf("Failed to read wsfFlags at %" PRIx64 "\n", addr + symbol_offsets.winsta_wsf_flags);
+        return VMI_FAILURE;
+    }
+
+    winsta->is_interactive = !(wsf_flags & 4);
+
+    addr_t desk = 0;
+
+    if (VMI_FAILURE == vmi_read_addr_va(vmi, addr + symbol_offsets.winsta_rpdesk_list_offset, 0, &desk))
+    {
+        printf("Failed to read pointer to rpdesklist at %" PRIx64 "\n", addr +
+            symbol_offsets.winsta_rpdesk_list_offset);
+        return VMI_FAILURE;
+    }
+
+    size_t len = 0x10;
+    winsta->desktops = (addr_t*)malloc(len * sizeof(addr_t));
+    memset(winsta->desktops, 0, sizeof(addr_t) * len);
+
```

Comment:
calloc ?

Reply:
Changed in 1e27e2979ac71a0dc3de2f3758ed61c48c15e3de